### PR TITLE
[11.0][MIG] auth_totp

### PR DESCRIPTION
**Pull request by MiquelRForgeFlow**
Standard migration.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Check Travis

**Comment by MiquelRForgeFlow** [comment]
@pedrobaeza GREEN :green_apple: :green_apple: :green_apple:

File: auth_totp/README.rst [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://odoo-community.org/logo.png>`_.
+
+Contributors
+------------
+
+* Oleg Bulkin <obulkin@laslabs.com>
```
Comment (simahawk) [inline auth_totp/README.rst:87]: Add "Do not contact contributors directly about support or help with technical issues."
See https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst

File: auth_totp/models/res_users_authenticator.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    )
+    user_id = fields.Many2one(
+        comodel_name='res.users',
+        ondelete='cascade',
+    )
+
+    @api.multi
+    @api.constrains('user_id')
+    def _check_has_user(self):
+        self.filtered(lambda r: not r.user_id).unlink()
```
Comment (simahawk) [inline auth_totp/models/res_users_authenticator.py:45]: a constrain that deletes?

File: auth_totp/wizards/res_users_authenticator_create.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+            totp = pyotp.TOTP(record.secret_key)
+            provisioning_uri = totp.provisioning_uri(
+                record.user_id.display_name.encode('utf-8'),
+                issuer_name=record.user_id.company_id.display_name,
+            )
+            provisioning_uri = urls.url_quote(provisioning_uri)
+
+            qr_width = qr_height = 300
+            tag_base = '<img src="/report/barcode/?type=QR&amp;'
```
Comment (simahawk) [inline auth_totp/wizards/res_users_authenticator_create.py:80]: not mandatory but I'd use a dict + urlencode

File: auth_totp/wizards/res_users_authenticator_create.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+class ResUsersAuthenticatorCreate(models.TransientModel):
+    _name = 'res.users.authenticator.create'
+    _description = 'MFA App/Device Creation Wizard'
+
+    name = fields.Char(
+        string='Authentication App/Device Name',
+        help='A name that will help you remember this authentication'
+             ' app/device',
+        required=True,
+        index=True,
```
Comment (simahawk) [inline auth_totp/wizards/res_users_authenticator_create.py]: why all these indexes? Isn't this a wizard? And I don't see any search :)

File: requirements.txt [hunk 5]
```diff
@@ -3,3 +3,4 @@ Pygments
 validate_email
 # auth_saml
 lasso
+pyotp
```
Comment (simahawk) [inline requirements.txt]: can you add the name of the module that requires it on top of it?

**Review by simahawk — CHANGES_REQUESTED** [review]

File: auth_totp/models/res_users_authenticator.py (hunk 2 above)
Comment (MiquelRForgeFlow) [reply auth_totp/models/res_users_authenticator.py:45]: That's from previous version. What should I do?

Comment (simahawk) [reply auth_totp/models/res_users_authenticator.py:45]: I dunno... :) How is it supposed to work? Who's resetting user_id? Or who's creating a record w/ no user_id?

**Review by JordiBForgeFlow — APPROVED** [review]
:+1: Functional review.

File: auth_totp/models/res_users_authenticator.py (hunk 2 above)
Comment (JordiBForgeFlow) [reply auth_totp/models/res_users_authenticator.py:45]: @obulkin Perhaps you know better. But to me, user_id should be a required field in 'res.users.authenticator'. Then this @api.constrains would not be required at all.

Comment (obulkin) [reply auth_totp/models/res_users_authenticator.py:45]: Existing authenticators are only exposed through an o2m field in a user view, which doesn't support deletion, so this logic is here to clean up orphaned authenticators when they're unmapped from a user via the view. You could make the field required and drop the `constrains`, but you would need to change the way that users manage their authenticators through the UI.

**Pull request merged**